File: source/com/barrybecker4/puzzle/tantrix1/ui/rendering/TantrixBoardRenderer.java
// Copyright by Barry G. Becker, 2012. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix1.ui.rendering;

import com.barrybecker4.common.geometry.Location;
import com.barrybecker4.puzzle.common.PuzzleRenderer;
import com.barrybecker4.puzzle.tantrix1.model.TantrixBoard;
import com.barrybecker4.puzzle.tantrix1.model.TilePlacement;

import java.awt.*;

import static com.barrybecker4.puzzle.tantrix1.ui.rendering.HexUtil.ROOT3;

/**
 * Renders the the tantrix puzzle onscreen.
 * @author Barry Becker
 */
public class TantrixBoardRenderer implements PuzzleRenderer<TantrixBoard> {

    private static final double MARGIN_FRAC = 0.2;
    static final int TOP_MARGIN = 15;

    private static final Color GRID_COLOR = new Color(130, 140, 170);

    private double hexRadius;
    private HexTileRenderer tileRenderer;

    /**
     * Constructor
     */
    public TantrixBoardRenderer() {
        tileRenderer = new HexTileRenderer();
    }

    /**
     * This renders the current state of the TantrixBoard to the screen.
     */
    public void render(Graphics g, TantrixBoard board, int width, int height)  {

        if (board == null) return;
        Graphics2D g2 = (Graphics2D) g;
        int minEdge = Math.min(width, height);
        hexRadius = (1.0 - MARGIN_FRAC) * minEdge / (board.getEdgeLength() * ROOT3 * .9);

        setHints(g2);
        drawGrid(g2, board);

        Location topLeftCorner = board.getBoundingBox().getTopLeftCorner();

        for (Location loc : board.getTantrixLocations()) {
            TilePlacement placement = board.getTilePlacement(loc);
            tileRenderer.render(g2, placement, topLeftCorner, hexRadius);
        }
    }

    private void setHints(Graphics2D g2) {
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING,
                RenderingHints.VALUE_ANTIALIAS_ON);
    }

    /**
     * Draw the gridlines over the background.
     */
    protected void drawGrid(Graphics2D g2, TantrixBoard board) {

        int edgeLen = board.getEdgeLength();
        int xpos, ypos;
        int i;
        int start = 0;
        int margin = (int)(hexRadius/2.0);
        double hexWidth = ROOT3 * hexRadius;
        int rightEdgePos = (int)(margin + hexWidth * edgeLen);
        int bottomEdgePos = (int)(TOP_MARGIN + margin + hexWidth * edgeLen);

        g2.setColor( GRID_COLOR );
        for ( i = start; i <= edgeLen; i++ )  //   -----
        {
            ypos = (int)(TOP_MARGIN + margin + i * hexWidth);
            g2.drawLine( margin, ypos, rightEdgePos, ypos );
        }
        for ( i = start; i <= edgeLen; i++ )  //   ||||
        {
            xpos = (int)(margin + i * hexWidth);
            g2.drawLine( xpos, TOP_MARGIN + margin, xpos, bottomEdgePos );
        }
    }
}

File: scala-source/com/barrybecker4/puzzle/tantrix/ui/rendering/TantrixBoardRenderer.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix.ui.rendering

import java.awt.*
import com.barrybecker4.common.geometry.{Box, IntLocation, Location}
import com.barrybecker4.puzzle.common.ui.PuzzleRenderer
import com.barrybecker4.puzzle.tantrix.model.{BoundingBoxCalculator, TantrixBoard, TilePlacement}
import com.barrybecker4.puzzle.tantrix.ui.rendering.HexUtil.ROOT3
import com.barrybecker4.puzzle.tantrix.ui.rendering.TantrixBoardRenderer.*
import scala.Iterable

/**
  * Renders the the tantrix puzzle onscreen.
  *
  * @author Barry Becker
  */
object TantrixBoardRenderer {
  private val MARGIN_FRAC = 0.2
  private[rendering] val TOP_MARGIN = 30
  private val GRID_COLOR = new Color(130, 140, 170)
  private val MIN_EDGE_LEN = 5
}

class TantrixBoardRenderer() extends PuzzleRenderer[TantrixBoard] {

  private val tileRenderer = new HexTileRenderer
  private var hexRadius = .0
  private var edgeLen: Int = _
  private val boundingBoxCalculator = BoundingBoxCalculator()

  def render(g: Graphics, board: TantrixBoard, width: Int, height: Int): Unit = {
    if (board != null)
      render(g, board.tantrix.tiles, width, height)
  }

  /**
    * This renders the current state of the TantrixBoard to the screen.
    */
  def render(g: Graphics, tiles: Iterable[TilePlacement], width: Int, height: Int): Unit = {
    if (tiles == null) return
    val g2 = g.asInstanceOf[Graphics2D]
    val bbox: Box = boundingBoxCalculator.getBoundingBox(tiles.toSeq)
    val boardEdgeLength = bbox.getMaxDimension
    val minEdge = Math.min(width, height)

    edgeLen = Math.max(MIN_EDGE_LEN, bbox.getMaxDimension)
    val padding = Math.max(0, MIN_EDGE_LEN - boardEdgeLength) / 2
    hexRadius = (1.0 - MARGIN_FRAC) * minEdge / (edgeLen * ROOT3 * .9)
    setHints(g2)

    val topLeftCorner = bbox.getTopLeftCorner.incrementOnCopy(-padding, -padding)
    drawGrid(g2, topLeftCorner)

    for (tile <- tiles) {
      tileRenderer.renderBorder(g2, tile, topLeftCorner, hexRadius)
    }
  }

  private def setHints(g2: Graphics2D): Unit = {
    g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
  }

  /**
    * Draw the gridlines over the background.
    */
  protected def drawGrid(g2: Graphics2D, topLeftCorner: Location): Unit = {
    //val margin = (hexRadius / 2.0).toInt
    //val hexWidth = ROOT3 * hexRadius
    //val rightEdgePos = (margin + hexWidth * edgeLen).toInt
    //val bottomEdgePos = (TOP_MARGIN + margin + hexWidth * edgeLen).toInt

    g2.setColor(GRID_COLOR)

    val bottomRightCorner = topLeftCorner.incrementOnCopy(edgeLen, edgeLen)
    for {
      i <- topLeftCorner.getY to bottomRightCorner.getY
      j <- topLeftCorner.getX to bottomRightCorner.getX
    } {
      tileRenderer.renderBorder(g2, IntLocation(i, j), topLeftCorner, hexRadius)
    }
  }
}
